Thunder Framework

davetcoleman opened the pull request:
> We present an experience-based planning framework
called Thunder that learns to reduce computation time
required to solve high-dimensional planning problems in varying
environments. The approach is especially suited for large
configuration spaces that include many invariant constraints,
such as those found with whole body humanoid motion planning.

I've tried to clean lots of stuff up, but in needs more run time testing and a demo for sure.

isucan commented on src/ompl/geometric/planners/experience/ThunderRetrieveRepair.h:
```diff
@@ -0,0 +1,205 @@
+             *  \brief Get debug information about the top recalled paths that were chosen for further filtering
+             *  \return chosenID - the index of the PlannerData object that was chosen for repair
+             */
+            const std::size_t& getLastRecalledNearestPathChosen() const;
```
PODs should not be returned by reference.

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/ThunderRetrieveRepair.h:
```diff
@@ -0,0 +1,205 @@
+             * \brief Get the chosen path used from database for repair
+             * \return PlannerData of chosen path
+             */
+            ompl::geometric::PathGeometric getChosenRecallPath() const;
```
No way to return a const & here? If not, maybe call this copyChosenRecallPath(geometric::PathGeometric *copy) ?

davetcoleman replied:
changed it

isucan commented on src/ompl/geometric/planners/experience/ThunderRetrieveRepair.h:
```diff
@@ -0,0 +1,205 @@
+             *  \brief Get debug information about the top recalled paths that were chosen for further filtering
+             *  \return data - vector of PlannerData objects that each hold a single path
+             */
+            const std::vector<ompl::geometric::PathGeometric>& getLastRecalledNearestPaths() const;
```
Here and elsewhere: the "ompl::" prefix is not needed.

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/ThunderRetrieveRepair.h:
```diff
@@ -0,0 +1,205 @@
+            /**
+             * \brief Pass a pointer of the database from the thunder framework
+             */
+            void setExperienceDB(ompl::tools::ThunderDBPtr experienceDB);
```
const & for passing the ptr in? That way we avoid a copy of the shared ptr

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/ThunderRetrieveRepair.h:
```diff
@@ -0,0 +1,205 @@
+             * \param oldPath - from experience
+             * \return true if no error
+             */
+            bool repairPath(ompl::geometric::PathGeometric &path, const base::PlannerTerminationCondition &ptc);
```
It is nice to have the args you write output data to as the last ones.

davetcoleman replied:
done. note i also had to change it in lightning framework

isucan commented on src/ompl/geometric/planners/experience/ThunderRetrieveRepair.h:
```diff
@@ -0,0 +1,205 @@
+            }
+
+            /** \brief Helper for outputing a state to console */
+            void debugState(const ompl::base::State* state)
```
printState already has std::cout as the default arg. This function does not seem very useful. Could we remove it?

davetcoleman replied:
good idea

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+ompl::geometric::ThunderRetrieveRepair::ThunderRetrieveRepair(const base::SpaceInformationPtr &si, const ompl::tools::ThunderDBPtr &experienceDB)
+    : base::Planner(si, "Thunder_Retrieve_Repair"),
+      experienceDB_(experienceDB),
+      nearestK_(10) // default value
```
make this a named magic constant (in the magic namespace)

davetcoleman replied:
done (used the same one from lightning)

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    {
+        // Set the repair planner
+        boost::shared_ptr<ompl::geometric::RRTConnect> repair_planner( new ompl::geometric::RRTConnect( si_ ) );
+        //repair_planner->setGoalBias(0.2);
```
remove comment

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    if (!repairPlanner_->isSetup())
+        repairPlanner_->setup();
+
+
```
remove white space?

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+        OMPL_INFORM("Experience database is empty so unable to run ThunderRetrieveRepair algorithm.");      
+        std::cout << ANSI_COLOR_RESET;
+
+        // TODO: it seems TIMEOUT causes the wrong behavior in parallel plan
```
How so? Could you open a bug?

davetcoleman replied:
I have no idea anymore. I changed comment to

> // Using CRASH because it seems TIMEOUT causes the wrong behavior in parallel plan, not sure why anymore

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    // Check if the database is empty
+    if (experienceDB_->isEmpty())
+    {
+        std::cout << ANSI_COLOR_GREEN;
```
Usual running of code in OMPL should produce no output at all to std::cout, other than that of the console STDHandler. Otherwise we can't redirect everything.

davetcoleman replied:
I've removed all colors

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    psk_->simplify(*primaryPath, ptc);
+    double simplifyTime = time::seconds(time::now() - simplifyStart);
+    OMPL_INFORM("ThunderRetrieveRepair: Path simplification took %f seconds and removed %d states", simplifyTime, numStates - primaryPath.getStateCount());
+    */
```
Does it make sense to do this? Make this an option of the algorithm, rather than have it as commented code?

davetcoleman replied:
Done - it defaults to off but can be turned on with ``enableSmoothing()``

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+
+bool ompl::geometric::ThunderRetrieveRepair::findBestPath(const base::State *startState, const base::State *goalState, ompl::base::PlannerDataPtr& chosenPath)
+{
+    return true;
```
Is this correct?

davetcoleman replied:
This was an unneeded function left over from lightning, good catch!

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    //        but that complicates everything and I'm not suppose to be spending too much time
+    //        on this prototype - DTC
+
+    std::cout << ANSI_COLOR_BLUE << std::endl;
```
Here ans elsewhere, please remove direct references to std::cout

davetcoleman replied:
Removed

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    // If not, replan between those states
+    for (std::size_t toID = 1; toID < primaryPath.getStateCount(); ++toID)
+    {
+        std::size_t fromID = toID-1; // this is our last known valid state
```
space around -

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+            {
+                OMPL_DEBUG("Before deleting invalid state part:");
+                primaryPath.print(std::cout);
+            }
```
remove?

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+                OMPL_INFORM("Deleting state %d", fromID + 1);
+                primaryPathStates.erase(primaryPathStates.begin() + fromID + 1);
+                --toID; // because vector has shrunk
+                OMPL_INFORM("toID is now %d", toID);
```
tmi?

davetcoleman replied:
removed second OMPL_INFORM

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+            }
+
+            if (superVerbose)
+            {
```
remove?

davetcoleman replied:
removed all superVerbose references

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+            //primaryPathStates.insert( primaryPathStates.begin() + toID, newPathSegment.getStates().begin(), newPathSegment.getStates().end() );
+            OMPL_DEBUG("Inserted new states into old path. New length: %d", primaryPathStates.size());
+            if (superVerbose)
+            {
```
remove superVerbose blocks?

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    return nearestPathsChosenID_; // of the candidate paths list, the one we chose
+}
+
+ompl::geometric::PathGeometric ompl::geometric::ThunderRetrieveRepair::getChosenRecallPath() const
```
const & pls

davetcoleman replied:
done

isucan commented on src/ompl/geometric/planners/experience/src/ThunderRetrieveRepair.cpp:
```diff
@@ -0,0 +1,456 @@
+    ompl::base::State *test = si_->allocState();
+
+    // Linerarly step through motion between state 0 to state 1
+    for (double location = 0.0; location <= 1.0; location += 1.0/double(segmentCount) )
```
precompute 1 / segmentCount rather than recomputing at every iteration?

davetcoleman replied:
done

isucan commented on src/ompl/tools/thunder/SPARSdb.h:
```diff
@@ -0,0 +1,802 @@
+           mechanics, but uses a different approach to identifying interfaces
+           and computing shortest paths through said interfaces.
+
+           This version has been modified for use with Thunder
```
Hmm. This makes me concerned. Can you describe the changes you make? Could we maybe add them directly to the planner as an option to avoid duplicating this file?

mamoll replied:
Or maybe derive from SPARS/SPARStwo?

davetcoleman replied:
I think this will be very involved to do right, and not worth the extra work (or feasible if I by chance make the next release). SPARS as implemented for OMPL has some differences that will not be easily combined. From the top of my head (its been a while):
- First SPARS builds roadmap from scratch for every problem
- A separate thread waits until the roadmap reaches a certain size/connectivity, then it attempts to plan on this roadmap while the roadmap is still being increased in size
- Once a path is found, SPARS stops growing the roadmap
- There is no easy way to save/load a roadmap from file in the SPARS implementation

In Thunder, I copied the SPARS file then made some modifications and called it SPARSdb. I removed many of the functions that were not necessary to me. I also implemented LazyPRM, so had to modify the graph node data a bit. Still, I am using the same connectivity checks that give us the advantageous theoretical guarantees

isucan commented on src/ompl/tools/thunder/ThunderDB.h:
```diff
@@ -0,0 +1,182 @@
+             * \param returned insertion time to add to db
+             * \return true on success
+             */
+            bool addPath(ompl::geometric::PathGeometric& solutionPath, double& insertionTime);
```
const solutionPath?

davetcoleman replied:
done

davetcoleman replied:
actually, it must be non-const because within this function it is interpolated

mamoll commented on src/ompl/tools/thunder/Thunder.h:
```diff
@@ -0,0 +1,210 @@
+            base::PlannerPtr                             rrPlanner_;
+
+            /**  A third planner used for testing dual-thread scratch-only planning */
+            base::PlannerPtr                             planner2_;
```
Is dual-thread scratch-only planning any different from using the ParallelPlan class?

davetcoleman replied:
No, except that I have my own experience-planning benchmarking functionality that no other planners in OMPL need (located in ExperienceSetup.h), and by having the dual planning functionality here it makes benchmarking much much easier

mamoll commented:
What is happening with this? Is there anything I can do to help get this merged?

davetcoleman commented:
I'm really sorry, but I'm not sure when I can find time to setup everything to do proper testing of this PR, while at my summer internship.

davetcoleman commented:
Ok, I've addressed all the comments, fixing almost all of them. I've rebased everything into two commits placed on top of the master branch. The second commit needs to be removed (its my debugging stuff - unless you would please please consider just adding these hooks) but I'll wait on removing it until you guys finish code reviewing.

davetcoleman commented:
Updated per Ioan's feedback today.

mamoll commented:
What has been updated?

davetcoleman commented:
Oh sorry, I've gotten into the habit of amending my commits to the previous commit. 

I added a base::PlannerStatus::ABORT type per [Ioan's suggestion](https://github.com/davetcoleman/ompl/commit/a1c9525a4463bf4910b1f3f824ede19fde13d1de#commitcomment-12480112)

I made a [function const](https://github.com/davetcoleman/ompl/commit/a1c9525a4463bf4910b1f3f824ede19fde13d1de#commitcomment-12480108)

I'll keep updated in a separate commit in the future

isucan commented:
Dave, I we can add a demo program, we can merge from my point of view

davetcoleman commented:
I've added both a Lightning and a Thunder demo that work.

mamoll commented:
I'd unify the two demos into one. It would not only minimize duplicated code, but also emphasize that these two planners are solving the same kinds of problems.
